apollo-federation/src/query_plan/fetch_dependency_graph.rs

```diff
@@ -443,45 +448,176 @@ impl ProcessingState {
+    pub(crate) fn new(
+        schema: ValidFederationSchema,
+        type_conditioned_fetching_enabled: bool,
+        root_type: CompositeTypeDefinitionPosition,
+    ) -> Self {
+        let root_possible_types: Vec<_> = if type_conditioned_fetching_enabled {
+            schema.possible_runtime_types(root_type).unwrap()
+        } else {
+            Default::default()
+        }
+        .into_iter()
+        .map(|pos| pos.get(schema.schema()).unwrap().clone())
+        .collect();
+
+        Self {
+            schema,
+            type_conditioned_fetching_enabled,
+            full_path: Default::default(),
+            path_in_node: Default::default(),
+            response_path: Default::default(),
+            possible_types: root_possible_types.clone(),
+            possible_types_after_last_field: root_possible_types,
+        }
+    }
     fn for_new_key_fetch(&self, new_context: Arc<OpPath>) -> Self {
         Self {
+            schema: self.schema.clone(),
             full_path: self.full_path.clone(),
             path_in_node: new_context,
             response_path: self.response_path.clone(),
+            type_conditioned_fetching_enabled: self.type_conditioned_fetching_enabled,
+            possible_types: self.possible_types.clone(),
+            possible_types_after_last_field: self.possible_types_after_last_field.clone(),
         }
     }
 
     fn add(
         &self,
         element: Arc<OpPathElement>,
     ) -> Result<FetchDependencyGraphNodePath, FederationError> {
+        let response_path = self.updated_response_path(&element)?;
+        let new_possible_types = self.new_possible_types(&element);
+        let possible_types_after_last_field = if let &OpPathElement::Field(_) = element.as_ref() {
+            new_possible_types.clone()
+        } else {
+            self.possible_types_after_last_field.clone()
+        };
+
         Ok(Self {
-            response_path: self.updated_response_path(&element)?,
+            schema: self.schema.clone(),
+            response_path,
             full_path: Arc::new(self.full_path.with_pushed(element.clone())),
             path_in_node: Arc::new(self.path_in_node.with_pushed(element)),
+            type_conditioned_fetching_enabled: self.type_conditioned_fetching_enabled,
+            possible_types: new_possible_types,
+            possible_types_after_last_field,
         })
     }
 
+    fn new_possible_types(&self, element: &OpPathElement) -> Vec<Node<ObjectType>> {
+        if !self.type_conditioned_fetching_enabled {
+            return Default::default();
+        }
+
+        match element {
+            OpPathElement::InlineFragment(f) => match &f.type_condition_position {
+                None => self.possible_types.clone(),
+                Some(tcp) => {
+                    let element_possible_types =
+                        self.schema.possible_runtime_types(tcp.clone()).unwrap();
+                    self.possible_types
```

Comment:
todo: make possible_types an indexset?

Reply:
Turns out it's not a great idea because we need ObjectTypes which don't implement hash/eq, so the cost to use IndexSet<Name> is going to be paid regardless.